src/wallet/rpcwallet.cpp

```diff
@@ -1304,6 +1306,237 @@ UniValue getshieldedbalance(const JSONRPCRequest& request)
+                + HelpExampleRpc("shielded_sendmany",
+                                 "\"DMJRSsuU9zfyrvxVaAEFQqK4MxZg6vgeS6\", [{\"address\": \"ps1ra969yfhvhp73rw5ak2xvtcm9fkuqsnmad7qln79mphhdrst3lwu9vvv03yuyqlh42p42st47qd\" ,\"amount\": 5.0}]")
+        );
+    EnsureWalletIsUnlocked();
+
+    // Check that the from address is valid.
+    auto fromaddress = request.params[0].get_str();
+    bool fromTaddr = false;
+    bool fromSapling = false;
+    CTxDestination fromTAddressDest = DecodeDestination(fromaddress);
+    fromTaddr = IsValidDestination(fromTAddressDest);
+    libzcash::SaplingPaymentAddress fromShieldedAddress;
+    if (!fromTaddr) {
+        auto res = KeyIO::DecodePaymentAddress(fromaddress);
+        if (!IsValidPaymentAddress(res)) {
+            // invalid
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid from address, should be a taddr or shielded addr.");
+        }
+
+        fromShieldedAddress = *boost::get<libzcash::SaplingPaymentAddress>(&res);
+
+        // Check that we have the spending key
+        if (!pwalletMain->HaveSpendingKeyForPaymentAddress(fromShieldedAddress)) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "From address does not belong to this node, shielded addr spending key not found.");
+        }
+
+        // Remember whether this is a Sprout or Sapling address
+        fromSapling = true;
+    }
+
+    UniValue outputs = request.params[1].get_array();
+    if (outputs.empty())
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, amounts array is empty.");
+
+    // Keep track of addresses to spot duplicates
+    std::set<std::string> setAddress;
+    // Recipients
+    std::vector<SendManyRecipient> taddrRecipients;
+    std::vector<SendManyRecipient> shieldAddrRecipients;
+    CAmount nTotalOut = 0;
+    bool containsSaplingOutput = false;
+
+    for (const UniValue& o : outputs.getValues()) {
+        if (!o.isObject())
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected object");
+
+        // sanity check, report error if unknown key-value pairs
+        for (const std::string& name_ : o.getKeys()) {
+            if (name_ != "address" && name_ != "amount" && name_!="memo")
+                throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("Invalid parameter, unknown key: ")+name_);
+        }
+
+        std::string address = find_value(o, "address").get_str();
+        bool isShieldAddr = false;
+        CTxDestination taddr = DecodeDestination(address);
+
+        if (!IsValidDestination(taddr)) {
+            auto res = KeyIO::DecodePaymentAddress(address);
+            if (IsValidPaymentAddress(res)) {
+                isShieldAddr = true;
+                bool toSapling = boost::get<libzcash::SaplingPaymentAddress>(&res) != nullptr;
```

Comment:
yeah, i left it here only because we have the sprout variant (it was part of the first PR #1553 integration when this was an on-going work). We don't have anything connected to it, matter of cleaning it up in a subsequent PR.

In a commit down the GUI line, i have directly implemented an DecodeSaplingAddress() returning an optional. So, going further, we can completely remove the `PaymentAddress` variant as well.

Reply:
Yes, absolutely.
What I meant here is that, even when `SproutPaymentAddress` is part of the `PaymentAddress` variant, it can never be returned by our implementation of `DecodePaymentAddress`, so  the `toSapling` variable here is redundant (always true).